Add PhpUnitTestAnnotationFixer

BackEndTea opened the pull request:
This PR turns php unit tests into either
```php
    /**
     * @test
     */
    public function seeIfItWorks() {...}
```
or 
```php
    public function testSeeIfItWorks() {...}
```
as mentioned in #3274 

Covered edge cases can be found in the tests.

keradus commented:
Hi, thanks for quickly picking it up ! Really great to see it here ;)
Remember, that this project is heavily maintained and for that we need to work with code that we are super sure about, thus, if you will get some picky comments to change sth or review process won't be that fast as all of us would like, please be understanding, and sorry for that ;)

Don't bother about `getPriority` here, it's indicator of "when" given fixer shall be executed in comparison to other fixers - before or after some of them. It's our way of configuring the order of conflicting fixers.
As I don't see any conflict here, you don't need to implement the method on your own (so, you will use default implementation that is in abstract parent class).

About definitions - you mean the failing test? I will point to it in comment of given line soon ;)

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,256 @@
+                new CodeSample("<?php
+class Test extends \\PhpUnit\\FrameWork\\TestCase
+{
+public function testItDoesSomething() {}}\n",['annotation' => true]),
```
the test is failing, as it expect to have `CodeSample` without `configuration` value, so one that uses default config.
this given `CodeSample` here is in fact the that one case, yet it has the config explicitly set up same as default one.

To solve issue fast, simply drop configuration from this `ConfigSample`.
To solve it nicely yet a bit more work, you could fix test to approve both, lack of explicit config and explicit config same as default one.

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,256 @@
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach (array_reverse($this->findPhpUnitClasses($tokens)) as $indexes) {
+            call_user_func([$this, $this->callback], $tokens, $indexes[0], $indexes[1]);
```
please avoid doing this. it's basically calling a function by string name, it's impossible to track that kind of usages by any Static Code Analysis tools. simply, make a condition and call different methods here explicitly

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,256 @@
+    protected function createConfigurationDefinition()
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('annotation', 'Whether to use the @test annotation or not.'))
```
It's great to have configuration. Tet, let us make it even more generic allowing ppl to add more possibilities in the future:
`style`, allowed values `['annotation', 'prefix']` with default as `prefix` (I do strongly believe from my experience that most ppl are actually using `prefix` over `annotation`, so please use it as default. yet, i'm also awared that some ppl like to use annotation, so this option will be great for them!)

keradus commented:
btw, please don't squash the commits.
like now, I see that sth has been modified, some of my comments are outdated now, but I have no way to see what was actually changed, so I need to start from beginning...

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,256 @@
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('annotation', 'Whether to use the @test annotation or not.'))
+            ->setAllowedValues([true, false])
```
consequtive methods shall be one level nested comparing to line with constructor

BackEndTea commented:
@keradus 
Will do, and thanks for the feedback so far

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,114 @@
+     * @param mixed $expected
+     * @param mixed $input
+     */
+    public function testFix($expected, $input, array $config)
```
usually, we do `array $config = []`, so you don't need to write `expected, null, []` but just `expected` in your first case

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,114 @@
+     * Case 2: Notation present, and not wanted
+     * Case 3: Notation not present, but wanted
+     * Case 4: Notation not present, but wanted, and there is already a notation
+     * Case 4: Notation not present, and not wanted
```
hint:
create data provider like
```php
return [
    'Notation present, and wanted' => [ ... ],
    'Notation present, and not wanted' => [ ... ],
];
```

and then your descriptions will be used by PHPUnit during reporting ;)

localheinz commented:
@BackEndTea 

In regard to priority, probably makes sense to run this before the `NoEmptyPhpdocFixer`, so if the only content of the docblock is the `@test` annotation, that it will eventually be removed.

👍 

By the way, great idea!

localheinz commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,256 @@
+    /**
+     * @test
+     */
+    public function itDoesSomething() {} }\n", ['annotation' => false])
```
@BackEndTea 

Just an idea (since you and I have seen this in at least one repository):

* some people use `camelCase` with `@test` annotation
* some people use `snake_case` with `@test` annotation
* some people use prefix `test` followed by `PascalCase`
* some maybe even use prefix `test` followed by `PascalCase` and also have a redundant `@test` annotation

Not sure if the fixer should take care of the last case (we could just add a separate fixer), but I think it would be awesome if the fixer could cater to all the 3 cases before, what do you think?

localheinz commented on README.rst:
```diff
@@ -1046,6 +1046,15 @@ Choose from the list of available rules:
+
+  Configuration options:
+
+  - ``annotation`` (``false``, ``true``): whether to use the @test annotation or not;
```
Quick heads-up, in case you wondered, `README.rst` will eventually need to be regenerated by running 

```
$ ./php-cs-fixer readme > README.rst
```

BackEndTea replied:
Ah thanks, forgot about that

keradus commented:
so.
before:
```php
/**
 * @test
 */
public function aaa() {}

/**
 * @test
 * @depends aaa
 */
public function bbb() {}
```

after:
so.

```php
/**
 */
public function testAaa() {}

/**
 * @depends aaa
 */
public function testBbb() {}
```

`@depends` annotation was not modified. tests will crash

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,260 @@
+            (new FixerOptionBuilder('style', 'Whether to use the @test annotation or not.'))
+                ->setAllowedValues(['prefix', 'annotation'])
+                ->setDefault('prefix')
+                ->setAllowedTypes(['string'])
```
no need to specify type when we only allowed explicitly pointed values.

BackEndTea commented:
Thanks for all the feedback so far.

@keradus 
I added a check for the ```@depends``` tags, and added tests to cover that behaviour.

@localheinz 
What priority do i have to give it in order to make sure that it is placed in the correct order?

And for your comment 
> * some people use camelCase with @test annotation
> * some people use snake_case with @test annotation
> * some people use prefix test followed by PascalCase
> * some maybe even use prefix test followed by PascalCase and also have a redundant @test annotation

This should cover case 1 and 3, but i'm not sure about how i should add support for case 2.

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,158 @@
+    public function itDoesSomething() {}
+    }', null, ['style' => 'annotation'],
+            ],
+            'Annotation is used, and it should not be' => ['<?php
```
let us globally change 
```php
$arr = ['aaa
bbb', 'ccc
ddd'];
```
into
```php
$aaa = [
    'aaa bbb',
    'ccc ddd',
];
```

we find it more readable

localheinz commented:
@BackEndTea @keradus 

I was just thinking, maybe it makes sense to separate this into two fixers?

* one that either removes the `test` prefix and adds the `@test` annotation or adds the `test` prefix and removes the `@test` annotation
* one that normalizes the test method name and is configurable, for example, accepts a `style` option
    * `camelCase`
    * `snake_case`
    * `wtf case` (see http://mnapoli.fr/using-non-breakable-spaces-in-test-method-names/)

Would that make sense?

localheinz commented:
@BackEndTea 

> What priority do i have to give it in order to make sure that it is placed in the correct order?

Maybe @keradus can help?

BackEndTea commented:
@localheinz 

I think i should be able to make this fix know if it should use camel or snake case, i'll get working on that soon. I don't mind adding support for wtf case, either, since it shouldn't be that hard, but i'm not sure if that is wanted.

One edge case i can think of is when the test goes from using an annotation to none, and the name, without test is just one word, but the standard for the project is snake case.

A fixer that normalises the test naming convention would be a separate fixer though.

keradus commented:
@localheinz

> I was just thinking, maybe it makes sense to separate this into two fixers?

in this PR, let us not make fixer that would fully normalize the test method.
Yet, if there is `/** @test */ function sth`, we need to know shall it be converted to `function testSth` or `function test_sth`.

> In the light of the latter, the former would probably just have to detect whether the test method name contains _, then prefix with test_, otherwise prefix with test and ucfirst() the current test method name (or something like that)
> Would that make sense?

No. When method is named with single word, like `pull`, you cannot detect it.


> > What priority do i have to give it in order to make sure that it is placed in the correct order?
> 
> Maybe @keradus can help?

out of the box - no clue. let us first finish spec of this fixer, then we will handle relations to different rules.


-----------

@BackEndTea

> I don't mind adding support for wtf case

no. don't. ever.


> One edge case i can think of is when the test goes from using an annotation to none, and the name, without test, is just one word, it wouldn't be able to know for sure what to use.

yeah, exactly. yet, we need to know which notation to use (camelCase vs snake_case).


> (Then again a test with a one word name is bad design and shouldn't happen)

No, it is not. There is a lot of valid cases to have single word.
(And even if it would be bad, we would still need to properly handle that)

> A fixer that normalises the test naming convention would be a separate fixer though.

exactly. let us not put it in scope of this PR.

BackEndTea commented:
Thanks for the insight @keradus 

> No. When method is named with single word, like pull, you cannot detect it.

I'm not sure how to properly detect it. And just defaulting to either camel or snake case would be bad. 

> > I don't mind adding support for wtf case
>
> no. don't. ever.

Got it.

>> A fixer that normalises the test naming convention would be a separate fixer though.
>
> exactly. let us not put it in scope of this PR.

Got it.

keradus commented:
> I'm not sure how to properly detect it.

you can't. you need config option for that.

BackEndTea commented:
I updated the addition logic to either add ```test_``` or ```test``` and uppercase the first letter of the original function, depending on the configuration.

For adding annotations, and thus removing ```test``` from the function name, it simply removes ```test_``` if its present, and otherwise it removes ```test```.

No wtf case support though.

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -127,8 +134,11 @@ private function applyTestAnnotation(Tokens $tokens, $startIndex, $endIndex)
-
-            $functionName = preg_replace('{test}', '', $functionName, 1);
+            if ($this->startsWith('test_', $functionName)) {
+                $functionName = preg_replace('{test_}', '', $functionName, 1);
```
why we need re here ?

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,335 @@
+                continue;
+            }
+            if ($this->startsWith('test_', $functionName)) {
+                $functionName = preg_replace('{test_}', '', $functionName, 1);
```
why we need re here ?

BackEndTea replied:
I updated it to use```substr``` instead now.

BackEndTea commented:
Known issue right now:
```php
    public function blabla() {}

    public function again() {}
```
becomes
```php
    /**
     * @test
     */
    public function blabla() {}
    /**
     * @test
     */

    public function again() {}
```

I can't seem to get it to properly add a new docblock if there is an empty new line above the function.

keradus commented:
> I can't seem to get it to properly add a new docblock if there is an empty new line above the function.

you need to split `T_WHITESPACE` into 2 tokens and add docblock between them

BackEndTea commented:
@keradus 
I have fixed that, and made sure the tests reflect that.

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+ */
+final class PhpUnitTestAnnotationFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface
+{
+    private $annotated;
```
please either typehint properties, or drop them (as we have `$configuration`)

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+        return $tokens[$docBlockIndex]->isGivenKind(T_DOC_COMMENT);
+    }
+
+    private function getDockBlockIndex(Tokens $tokens, $index)
```
typehinting is missing

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+     *
+     * @return bool
+     */
+    private function shouldBeChecked(Tokens $tokens, $index)
```
should **what** be checked ?

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+     *
+     * @return Line[]
+     */
+    private function updateDocBlock(Tokens $tokens, $docBlockIndex)
```
this looks overcomplicated. what is the purpose of this method ?

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+    public function a_helper_function () {}
+
+    /**
+     * @depends depends_on_someone
```
please add test case with more complicated docblock, like
```php
/**
 * Description.
 *
 * @param ...
 *
 * @return ...
 *
 * @depends ...
 */
```

keradus replied:
here, in input code, the method we depends on is still called `test_depends_on_someone`.
It shall remain `depends_on_someone` in expected code

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+    public function helper_function() {}
+}',
+                ['case' => 'snake'],
+                ],
```
wrong indent

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+    public function test_works_fine_too() {}
+}',
+                    ['style' => 'annotation'],
+                    ],
```
wrong indent

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+}',
+                    ['case' => 'snake'],
+                ],
+            'Class has both camel and snake case, annotated functions and not, and wants to add annotations' => [
```
again...

keradus replied:
this is pretty nice test case I was thinking about, big :+1: you added it already :)

yet, let us extend a bit, as I found edge case that is not properly handled
https://github.com/BackEndTea/PHP-CS-Fixer/pull/1

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+    public function testCamelCased () {}
+
+    /**
+     * @depends testCamelCased
```
please add test case with more complicated docblock, like
```php
/**
 * Description.
 *
 * @param ...
 *
 * @return ...
 */
```

keradus replied:
it's added

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+     * @dataProvider provideFixCases
+     *
+     * @param mixed $expected
+     * @param mixed $input
```
it shouldn't be mixed

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+    /**
+     * @test
+     */
+    public function itDoesSomething() {} }\n"),
```
please don't use `\n` inside `"` strings. just use real new line.
also, please put closing `}` of class on proper, dedicated line

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+        $originalIndent = $this->detectIndent($tokens, $tokens->getNextNonWhitespace($docBlockIndex));
+        $toInsert = [
+            new Token([T_DOC_COMMENT, "/**\n$originalIndent * @test\n$originalIndent */"]),
+            new Token([T_WHITESPACE, "\n".$originalIndent]),
```
you need to use whitespace config, and then use proper line ending / indentation.
for now, it will produce wrong code for ppl using tabs / windows line endings.

cover such cases in new test method called `testMessyWhitespaces` (grep for it to get idea what to do)

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -0,0 +1,427 @@
+     * @test
+     */
+    public function itDoesSomething() {}
+    }',
```
wrong indent

keradus commented:
could you please include https://github.com/BackEndTea/PHP-CS-Fixer/pull/2 ? ;)

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+
+    /**
+     * @param Tokens $tokens
+     * @param $index
```
type is missing

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+
+    /**
+     * @param Tokens $tokens
+     * @param $index
```
type is missing

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+
+    /**
+     * @param Tokens $tokens
+     * @param $docBlockIndex
```
type is missing

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,375 @@
+
+    /**
+     * @param Tokens $tokens
+     * @param $docBlockIndex
```
type is missing

keradus commented:
@BackEndTea, thanks for great work!
I raised some comments, mainly minors, yet one bug and lack of whitespaces configuration are the blockers

BackEndTea commented:
@keradus I fixed the minor issues, as well as the bug, and the whitespace configuration.

BackEndTea commented:
I noticed an error, which also isn't covered by any test cases. It currently breaks when trying to handle a test that was annotated like this ```/** @test */``` instead of
```php
/**
 * @test
 */
```

BackEndTea commented:
On that note, i figured out some more edge cases involving single line doc blocks that aren't covered yet.

keradus commented:
this one-liner -> multiliner annotation is done only when annotation is added, right?

about dedicated fixer to do it always, let us finish this PR first and then we will see ;)


Sounds really great.
Sadly, my time this week is very limited. Let me try to do next round next week, or perhaps someone else is willing to take a look as well ? :)

BackEndTea commented:
Yeah the one liner to multi liner is only done when adding the ```@test``` annotation.

I could also change it to convert ```/** @group Database */``` to ```/** @test @group Database */```, instead of
```
/**
 * @test
 * @group Database
 */
```

But thats up to the maintainers to decide

keradus commented:
nah, when adding annotation convert to multiliner, that makes sense :+1:

BackEndTea commented:
I'm a little bit stuck on something
```php
/**
 * asdf@testasdf
 */
```
Gets accepted by phpUnit just the same as
```php
/**
 * @test
 */
```

Should this fixer 'see' the ```@test``` annotation in the first example?

keradus commented:
> I'm a little bit stuck on something (...) Gets accepted by phpUnit (...)

Don't bother about it.
IMO, you shall open a bug report ;)

> Should this fixer 'see' the `@test` annotation in the first example?

I believe that our DocBlock parser is not see that annotation, and it's good if it does't.
If that's the case, you have nothing to do with it.

> Of course it would remove the test prefix, but should it add an explicit @test to the Doc block?

yes, it should.

BackEndTea commented:
Things I have no idea how to currently do/fix:

- How to handle abstract test functions
- How to handle interfaces with test functions
- How to handle traits with test functions

For example:
```php
trait ItsAUser 
{
    public function testItsAUser()
    {
        $this->assertTrue($this->user->isUser());
    }
}

class TestUser extends \PHPUnit\Framework\TestCase
{
    use ItsAUser;
    
    protected $user;
    
    protected function setUp()
    {
        $this->user = new \Foo\User();
        
    }
}
```

There is good way to check that the trait is actually a test, and its not just a normal function that starts with test.

keradus commented on tests/Fixer/PhpUnit/PhpUnitTestAnnotationFixerTest.php:
```diff
@@ -931,14 +899,93 @@ class Test extends \PhpUnit\FrameWork\TestCase
+     * @param string $input
+     * @param array  $config
+     */
+    public function testItDoesNotChange($input, array $config = [])
```
please move it back to one test method / data provider

BackEndTea replied:
Okay

keradus commented:
> How to handle abstract test functions

like any other tets function

> How to handle interfaces with test functions
> How to handle traits with test functions

for now, ignore them. let us focus on finishing PR rather than extending it's scope, it could be done later if needed

BackEndTea commented:
@keradus 
My concern is this, I could have an abstract function lets say 
```php
    abstract public function itDoesThings();
```
Now, when implemented, it is done like this
```php
    /**
     * @test
     */
    public function itDoesThings() {}
```

In the abstract class, there is no way to guarantee that an abstract function is a test or not

keradus commented:
then in abstract class it shall be marked with `@test` as well.

look, when one use other way to mark method as a test - a prefix, then he mark that abstract method with prefix as well.


> But it could break tests if the fixer ran over this.

any change of function name can break code. very good point. note that all of our phpunit fixers are marked as risky.

yes, mark this fixer as risky as well.

BackEndTea commented:
>yes, mark this fixer as risky as well.

Will do

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -0,0 +1,572 @@
+     *
+     * @return bool
+     */
+    private function doesDocBlockContainTest(DocBlock $doc)
```
can we just get annotation from docblock? it has all the logic already

keradus replied:
ping @BackEndTea

BackEndTea replied:
I'm sorry, i don't understand exactly what it is you are asking.

keradus replied:
```php
private function doesDocBlockContainTest(DocBlock $doc)
{
    return !empty($doc->getAnnotationsOfType('test'));
}
```

BackEndTea replied:
Then we get some issues with functions annotated like this:
```php
/**
 * This is a @test
 */
```
Which is a valid phpunit test

BackEndTea replied:
It causes the ```Annotation does not get added when it is already present in a weird place``` case to break.

If we don't want to support that, i'm okay with that too

keradus replied:
as said in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3275#issuecomment-349796720 , we shouldn't handle that scenario

BackEndTea replied:
👍 I'll update it then

keradus replied:
done

keradus commented:
I let myself fix last raised issue

keradus commented:
new concerns from final tests:

1. when we want to use `@test` annotation and current test method name is `test123foo`, fixer would rename method to `123foo`, which is PHP syntax issue, this has to be fixed (probably adding annotation, yet leaving test prefix)

2. fixer is mature enough, let us fix priority:
fixer must run before `phpdoc_trim` and `no_empty_phpdoc`

3. again for `@test` anontation over prefix.
`testGetName` is renamed to `getName`. it's the issue, as it's overwriting internal PHPUnit method (most probably with wrong params, as test could have no params, and internal one has a param). Not sure about best approach here.

BackEndTea commented:
I added a test, and a fix for issue 1 you raised.

BackEndTea commented:
The only way around issue 3 i can really see is having a list of all phpunit functions, and adding those to some kind of black list. But that is going to be a big list

keradus commented on src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php:
```diff
@@ -65,6 +65,12 @@ public function testItDoesSomething() {}}'.$this->whitespacesConfig->getLineEndi
 
+    public function getPriority()
+    {
+        // must be run before the PhpdocSeparationFixer and PhpdocOrderFixer
```
you need to also add this as a test here:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.8/tests/AutoReview/FixerFactoryTest.php#L182

in tests here, you also have a comments with filename of integration test. you will need to add 2 of them, take a look at other ones so you will get idea how to create them

BackEndTea replied:
I'm adding those now, as well as running the fixer again

keradus commented:
> But that is going to be a big list

and imposible to maintain whenever external phpunit class will change/add/remove it's methods

keradus commented:
please freeze now

BackEndTea commented:
:+1

keradus commented:
damn, nightly was failing...

keradus commented:
Thank you @BackEndTea.

keradus commented:
really great work here !

BackEndTea commented:
Real big thanks on all the feedback and help @keradus & @localheinz
